internal/provider/resource_access_model_test.go

```diff
@@ -147,3 +147,91 @@ resource "juju_access_model" "test" {
   users = [juju_user.test-user.name]
 }`, userName, userPassword, modelName, access)
 }
+
+func TestAcc_ResourceAccessModel_Schema_v0_To_v1(t *testing.T) {
+	SkipJAAS(t)
+	user1 := acctest.RandomWithPrefix("tfuser")
+	password1 := acctest.RandomWithPrefix("tf-test-user")
+	user2 := acctest.RandomWithPrefix("tfuser")
+	password2 := acctest.RandomWithPrefix("tf-test-user")
+	modelName := acctest.RandomWithPrefix("tf-access-model")
+
+	resourceName := "juju_access_model.test"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() { testAccPreCheck(t) },
+		Steps: []resource.TestStep{
+			{
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"juju": {
+						VersionConstraint: "0.21.1",
+						Source:            "juju/juju",
+					},
+				},
+				Config: testAccResourceAccessModelTwoUsers(user1, password1, user2, password2, modelName, "write", false),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "model", modelName),
+					resource.TestCheckResourceAttr(resourceName, "access", "write"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user1),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user2),
+				),
+			},
+			{
+				ProtoV6ProviderFactories: frameworkProviderFactories,
+				Config:                   testAccResourceAccessModelTwoUsers(user1, password1, user2, password2, modelName, "write", false),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "model", modelName),
+					resource.TestCheckResourceAttr(resourceName, "access", "write"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user1),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user2),
+				),
+			},
+			{
+				ProtoV6ProviderFactories: frameworkProviderFactories,
+				Config:                   testAccResourceAccessModelTwoUsers(user1, password1, user2, password2, modelName, "write", true),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "model", modelName),
+					resource.TestCheckResourceAttr(resourceName, "access", "write"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user1),
+					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", user2),
```

Comment:
This test step is hard to read at a glance. I'd add a comment above it if you choose to keep it.

Also, I don't **think** it's actually verifying what you want. If we were using a `list` this test step would remove access then re-add access and we wouldn't tell from the test. Instead you want to add `PlanOnly: true`, the docstring for that field says,
```
	// PlanOnly can be set to only run `plan` with this configuration, and not
	// actually apply it. This is useful for ensuring config changes result in
	// no-op plans
	PlanOnly bool
```